### PR TITLE
fix(core): enforce required media and relations via api.documents.strictRelations

### LINKED ISSUE #24927: Media field can be null if requiered
### Node Version

22.14.0

### NPM/Yarn/PNPM Version

10.9.2

### Strapi Version

5.23.4

### Operating System

Windows 10

### Database

PostgreSQL

### Javascript or Typescript

Typescript

### Reproduction URL

_No response_

### Bug Description

When i create a media field and set it to requiered i can still create a entry without needing to apply a media this is for multiple media and for single media, also for multiple media it returns null if no image selected instead of an empty array as expected

### Steps to Reproduce

Single Media 
1. Create Content Type with field image and set the field to a Media field
2. Set it to requiered
3. Create a new entry and see it doesnt enforce a media (single media) -> returning null instead

Multiple Media 
1. Create Content Type with field image and set the field to a Media field
2. Set it to requiered
3. Create a new entry without images
4. Call the rest api and see it returns null instead of an empty array

### Expected Behavior

For single media if requiered it should enforce that a image is selected

And for multiple media it should not return null if 0 images are selected instead return an empty array

### Logs

```shell

```

### Code Snippets

_No response_

### Media

_No response_

### Additional information

_No response_

### Confirmation Checklist

- [ ] I have checked the existing [issues](https://github.com/strapi/strapi/issues) for duplicates.
- [x] I agree to follow this project's [Code of Conduct](https://github.com/strapi/strapi/blob/develop/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by giu1io**
### What does it do?

It adds a new config option, `api.documents.strictRelations`. When it's on, Strapi actually checks that required media and required relation fields are filled in before an entry is published.

* New projects get it turned **on** by default (same as the existing `strictParams` option).
* Existing projects are **not affected** — if the option isn't set, nothing changes.
* Drafts are still allowed to be empty. The check only runs when you publish (or write directly to a non-draft entry). This matters because it means you can still upload a file while creating an entry — that used to break Strapi back in 2023 ([#14670](<https://github.com/strapi/strapi/issues/14670>) → [#16002](<https://github.com/strapi/strapi/issues/16002>)), and this avoids repeating that mistake.
* A relation value can be sent in a few different shapes (a single id, a list of ids, or a `{ connect }` / `{ set }` object). The check understands all of them, so it works the same whether you write directly or publish.

What gets checked in this version: a required single media, a required multi-media field (must have at least one), and required to-one / to-many relations.

**Left for later:** empty multi-media fields still come back as `null` instead of `[]` (the second half of strapi/strapi#24927). Fixing that touches how data is read back and is riskier, so it'll be a separate PR. This PR only handles the "required must not be empty" part.

### Why is it needed?

Right now, a field marked `required: true` isn't really enforced for media and relations. You can create **and publish** an entry with a required media or relation left empty, and Strapi won't complain.

We can't just fix it directly, because that would suddenly reject writes that used to work, and a past attempt at fixing it broke file uploads during entry creation ([#16002](<https://github.com/strapi/strapi/issues/16002>)). So instead we put the fix behind an option that's on by default for new projects and off for everyone else.

### How to test it?

```bash
yarn test:generate-app
yarn test:api --generate-app=false \
  tests/api/core/content-manager/media/required-media.test.api.js \
  tests/api/core/content-manager/media/required-relation.test.api.js
yarn test:unit packages/core/core
```

By hand: in a project with `documents.strictRelations: true` in `config/api`, make a content type with a required media (or relation), leave it empty, and save the draft — that works. Then try to publish it — you get a `400`. Add the file/relation and publish again — it works. With the option off, nothing changes.

### Related issue(s)/PR(s)

Fix strapi/strapi#24927. Also helps with [#14648](<https://github.com/strapi/strapi/issues/14648>), [#15445](<https://github.com/strapi/strapi/issues/15445>), [#19433](<https://github.com/strapi/strapi/issues/19433>). Background on the earlier revert: [#14670](<https://github.com/strapi/strapi/issues/14670>) → [#16002](<https://github.com/strapi/strapi/issues/16002>). A follow-up PR will handle the `null` vs `[]` part.

🤖 Generated with [Claude Code](<https://claude.com/claude-code>)

### CODE REVIEW

File: packages/core/core/src/services/entity-validator/index.ts [hunk 1]
```diff
@@ -235,17 +236,87 @@ const createDzValidator =
...
+  // via `hasRelationValue`, which understands every shape.
+  let validator: strapiUtils.yup.AnySchema = yup.mixed();
+
+  // Relational required constraints are only enforced under the strictRelations flag,
+  // and never for drafts (mirrors scalar `required` handling via `!isDraft`).
+  if (options.strictRelations && !options.isDraft && attr.required) {
+    // A media value can arrive as an id, an array, or a connect/set object, so a simple
+    // `.notNil()`/`.min()` is not enough — normalise the shape before asserting.
+    validator = validator.test('required-media', `${updatedAttribute.name} must be defined.`, () =>
+      hasRelationValue(updatedAttribute.value)
```
Comment (innerdvations) [inline packages/core/core/src/services/entity-validator/index.ts]: Partial non-draft updates that omit a required media/relation currently fail validation (`… must be defined`) even when the entity already has a value.

Scalar `required` on update allows absent keys via `notNull()`; `strictRelations` should mirror that and only reject when the field is **present** and empty (`null` / `[]` / empty `set`/`connect`). This matters for non-DP types, where new apps get `strictRelations: true` by default.

Same pattern applies in `createRelationValidator` below.

_AI-assisted comment._

File: packages/core/core/src/services/entity-validator/index.ts [hunk 2]
```diff
@@ -332,10 +409,13 @@ const createAttributeValidator =
```
Comment (innerdvations) [inline packages/core/core/src/services/entity-validator/index.ts:512]: `createComponentValidator` / `createDzValidator` forward only `{ isDraft }` into nested `createModelValidator` (see the `{ isDraft }` args inside those helpers), so `strictRelations` is dropped here even though `options` is passed into this call.

Required media/relations inside components or dynamic zones are never enforced. Please pass the full validator options (at least `{ isDraft, strictRelations, locale }`) through both recursion sites.

_AI-assisted comment._

File: packages/cli/create-strapi-app/templates/vanilla/config/api.ts [hunk 3]
```diff
@@ -9,6 +9,7 @@ const config: Core.Config.Api = {
   },
   documents: {
     strictParams: true,
+    strictRelations: true,
```
Comment (innerdvations) [inline packages/cli/create-strapi-app/templates/vanilla/config/api.ts:12]: Templates now set `documents.strictRelations: true`, but [Recommended security defaults](https://github.com/strapi/strapi/blob/develop/docs/docs/docs/01-core/configuration/01-security-defaults.md) still only documents `strictParams`. Please add `strictRelations` to that snippet so scaffolding and contributor docs stay aligned.

_AI-assisted comment._

**Review by innerdvations — CHANGES_REQUESTED** [review]
Thanks for the careful opt-in design (drafts lenient, new apps default on, existing projects untouched). Requesting changes for the two production bugs called out inline; two smaller items below.

### Also (non-blocking)

**Content Releases validity check** — `isEntryValid` in `packages/core/content-releases/server/src/utils/index.ts` calls `validateEntityCreation` with `undefined` options, so it never reads `api.documents.strictRelations`. Release UI can mark an entry valid that publish will later reject under the flag. Publish itself is gated via document-service; this is a validity-check consistency gap.

_AI-assisted comment._

File: packages/core/core/src/services/entity-validator/index.ts (hunk 1 above)
Comment (giu1io) [reply packages/core/core/src/services/entity-validator/index.ts]: Fixed in `fix(core): correct strictRelations update semantics and nested enforcement` — the media/relation validators are now curried on `createOrUpdate` and mirror scalar semantics:

- **creation** (`notNil` analogue): reject any empty value (absent / `null` / `[]` / empty `set`/`connect`).
- **update** (`notNull` analogue): an absent key keeps the existing value and passes; only reject when the request is present and provably empties the field.

One deviation from the review wording worth confirming: a bare `{ connect: [] }` (no `set`) on **update** is treated as a **no-op that passes**, not a reject. It adds nothing and removes nothing, so it can't be what empties a required field — and the admin CM can emit it when diffing an unchanged relation, so rejecting would be a false positive. The absolute shapes (`null`, `[]`, `{ set: [] }`) still reject on update; `{ connect: [] }` on **creation** still fails. The rule I implemented is *"does this request provably leave the field empty?"* rather than *"is the payload literally empty?"*. Good to keep it this way?

Pinned with unit tests in `strict-relations-validators.test.ts`: `{ connect: [] }` passes on update, throws on creation, and `{ connect: [id] }` passes — so the no-op carve-out is scoped to the empty case only.

Comment (giu1io) [reply packages/core/core/src/services/entity-validator/index.ts]: Related limitation on the same update path, worth calling out explicitly: a disconnect-only update like `{ disconnect: [5] }` is **not** validated at write time. Whether it empties a required relation depends on current DB state the entity validator doesn't load, so it falls through to a pass.

That said, it can't produce an invalid *published* entity. `publish()` reloads the draft with `getDeepPopulate` and re-runs `validateEntityCreation` with `isDraft: false` against the materialised state (`repository.ts` `publish` → `entries.ts` `publishEntry` → `createEntry`), so a required relation emptied via disconnect is rejected at publish with a 400. The write-time check is a fast-fail UX layer; publish is the authoritative gate. (This is the same populated-state check content-releases' `isEntryValid` now uses too, after this PR.)

So this is a write-time-only gap, not a data-integrity one. OK to treat as a documented v1 limitation, tracked with the `null → []` read-back follow-up?

Documented with tests: a unit test in `strict-relations-validators.test.ts` asserts `{ disconnect: [id] }` passes at write time (relation and media), and a new integration test `strict-relations-publish.test.api.js` proves the backstop — disconnecting the last required relation updates the draft (200) but publish returns 400, while a no-op `{ connect: [] }` and a disconnect+reconnect both still publish (200).

File: packages/core/core/src/services/entity-validator/index.ts (hunk 2 above)
Comment (giu1io) [reply packages/core/core/src/services/entity-validator/index.ts:512]: Fixed — `createComponentValidator` and `createDzValidator` now take the full `options` object and forward it into every nested `createModelValidator` call, instead of reconstructing `{ isDraft }`. So `strictRelations` (and `locale`) survive the recursion and required media/relations nested in components and dynamic zones are enforced.

Covered by unit tests (single + repeatable component, and DZ: enforced when strict + non-draft, lenient for drafts, off when the flag is unset) and API tests in `required-relation.test.api.js` (publishing a component / DZ with an empty required nested relation → 400, populated → 200).

File: packages/cli/create-strapi-app/templates/vanilla/config/api.ts (hunk 3 above)
Comment (giu1io) [reply packages/cli/create-strapi-app/templates/vanilla/config/api.ts:12]: Fixed — added `documents.strictRelations: true` to the snippet in [Recommended security defaults](https://github.com/strapi/strapi/blob/develop/docs/docs/docs/01-core/configuration/01-security-defaults.md) plus a short "Strict required relations" section explaining that it enforces `required` on media/relations for published (non-draft) writes, applies to nested components/DZs, and is on by default in new apps only (existing projects keep the lenient behaviour until they opt in).

**Comment by giu1io** [comment]
**Content Releases validity check (non-blocking item from the review summary)** — fixed in `fix(content-releases): read strictRelations in isEntryValid`. `isEntryValid` now reads `api.documents.strictRelations` and passes `{ isDraft: false, locale, strictRelations }` to `validateEntityCreation` (previously `undefined`). The entry is already deep-populated (`populateDeep(Infinity)`), so it validates against the same materialised state the publish action does — the release "valid" badge and what publish accepts are now consistent. Covered by a new unit test `content-releases/server/src/utils/__tests__/is-entry-valid.test.ts`.

File: packages/core/core/src/services/entity-validator/index.ts (hunk 1 above)
Comment (innerdvations) [reply packages/core/core/src/services/entity-validator/index.ts]: Thanks for the follow-ups — the earlier findings look fixed (update omit semantics, nested options forwarding, security-defaults docs, and content-releases `isEntryValid`). The `{ connect: [] }` no-op on update is fine to keep.

One remaining gap: with `strictRelations` on, a non-draft update can still empty a required media/relation via `{ disconnect: […] }` (no `connect`/`set`) and succeed — same empty end state, different payload shape. For DP, publish still catches it; for non-DP there is no later gate, so the empty required field can stick. Worth closing before merge — at least reject disconnect-only on required to-one / single media on non-draft writes (those always empty without a DB read).

_AI-assisted comment._

**Comment by giu1io** [comment]
@innerdvations — before I commit these, wanted to check they're worth adding.

Following up on the connect/disconnect edge cases from the update-semantics thread, I wrote an integration suite (`tests/api/core/content-manager/media/strict-relations-publish.test.api.js`) that exercises the **publish-time backstop** end-to-end under `strictRelations: true` — i.e. the cases write-time validation intentionally lets through but publish must catch when re-validating the deep-populated draft as a non-draft creation:

**Relations (oneToOne):**
- disconnect the last required relation → update 200, draft empty, **publish 400**
- no-op `{ connect: [] }` on a populated draft → preserved, **publish 200**
- disconnect then reconnect → populated, **publish 200**

**Single required media (real uploaded file):**
- disconnect the last file → update 200, draft empty, **publish 400**
- `{ connect: [] }` → **empties the media** → **publish 400** (see divergence below)
- disconnect then reconnect → populated, **publish 200**

Two questions:

1. **Are these worth keeping in the PR?** They overlap somewhat with the existing `required-media` / `required-relation` publish tests, but they specifically cover the connect/disconnect *deltas* and prove publish is the authoritative gate for the disconnect-only write-time gap we discussed. Happy to trim to just the relation disconnect→400 case, keep the full set, or drop them if you think the unit coverage in `strict-relations-validators.test.ts` is enough.

2. **Divergence I hit while writing them (not introduced by this PR):** on a **single media**, `{ connect: [] }` and `{ disconnect: [] }` *empty* the field — single media treats an object-form payload as a full replace. On a **oneToOne relation**, `{ connect: [] }` is a no-op that preserves the value. It reproduces with the flag off and with an optional field, and the emptying happens at the DB write (before any validation), so it's pre-existing media behaviour, orthogonal to strictRelations. I left production code untouched and made the media test assert the actual behaviour, with a comment explaining it. Do you want this tracked as a separate issue (should single media honour connect/disconnect deltas like relations?), or is it expected/known?

**Comment by innerdvations** [comment]
> 1. **Are these worth keeping in the PR?** They overlap somewhat with the existing `required-media` / `required-relation` publish tests, but they specifically cover the connect/disconnect _deltas_ and prove publish is the authoritative gate for the disconnect-only write-time gap we discussed. Happy to trim to just the relation disconnect→400 case, keep the full set, or drop them if you think the unit coverage in `strict-relations-validators.test.ts` is enough.

Yeah, I think they're still useful and prove the full path, a bit of overlap is fine.

> 2. **Divergence I hit while writing them (not introduced by this PR):** on a **single media**, `{ connect: [] }` and `{ disconnect: [] }` _empty_ the field — single media treats an object-form payload as a full replace. On a **oneToOne relation**, `{ connect: [] }` is a no-op that preserves the value. It reproduces with the flag off and with an optional field, and the emptying happens at the DB write (before any validation), so it's pre-existing media behaviour, orthogonal to strictRelations. I left production code untouched and made the media test assert the actual behaviour, with a comment explaining it. Do you want this tracked as a separate issue (should single media honour connect/disconnect deltas like relations?), or is it expected/known?

Yeah, worth a separate issue so that we can follow-up on it or document it.

File: packages/core/core/src/services/entity-validator/index.ts (hunk 1 above)
Comment (giu1io) [reply packages/core/core/src/services/entity-validator/index.ts]: Ok should be addressed now, if you can do (hopefully) one last check 🙏

**Review by jhoward1994 — DISMISSED** [review]

**Comment by scs-eng** [comment]
Thanks for working on this!
Will you also take into account the fact that Strapi regenerates schemas after content-type changes and removes any manually added "required": true option from relation fields?

**Comment by giu1io** [comment]
> Thanks for working on this! Will you also take into account the fact that Strapi regenerates schemas after content-type changes and removes any manually added "required": true option from relation fields?

Good question @scs-eng — but that's out of scope for this PR. The main focus here was to fix the issue with the required media fields. More broadly, this change is about enforcement: when `api.documents.strictRelations` is on, Strapi actually validates that a required media/relation is filled in on publish. It relies on required: true already being present in the schema (set via `schema.json`).

The Content-Type Builder side — not exposing a required toggle for relations in the UI, and regenerating schemas in a way that drops a manually-added required — is a separate, long-standing topic tracked in #15445.

File: packages/core/core/src/services/entity-validator/index.ts (hunk 1 above)
Comment (innerdvations) [reply packages/core/core/src/services/entity-validator/index.ts]: Update-side to-one / single-media / CM shapes look good — including the untouched `{ connect: [], disconnect: [] }` no-op and the non-D&P API cases. One more hole on the **creation** path: `hasRelationValue` treats a bare `{ disconnect: [id] }` (no `connect`/`set`) as a present value, so `validateEntityCreation` with `strictRelations: true` accepts a required to-one that attaches nothing. Please reject disconnect-only on creation too (to-one and to-many — always empty on create). I have a failing unit case locally you can lift.

_AI-assisted comment._

Comment (giu1io) [reply packages/core/core/src/services/entity-validator/index.ts]: @innerdvations ok fix pushed. Hopefully we are good now 😅 thank you 🙏

**Review by innerdvations — APPROVED** [review]
Tested and LGTM!
